File: python/src/cccc_sdk/client_0430_admin_ops.py
from __future__ import annotations

from typing import Any, Dict, Optional, Union

from .client_0430_shared import _compact
from .errors import DaemonAPIError


_MAX_GROUP_PREAMBLE_BYTES = 512 * 1024


class CCCC0430AdminOpsMixin:
    def actor_new_session(
        self,
        *,
        group_id: str,
        actor_id: str,
        by: str = "user",
    ) -> Dict[str, Any]:
        return self.call(
            "actor_new_session",
            {
                "group_id": str(group_id),
                "actor_id": str(actor_id),
                "by": str(by),
            },
        )

    def group_copy_export_file(self, *, group_id: str, by: str = "user") -> Dict[str, Any]:
        return self.call("group_copy_export_file", {"group_id": str(group_id), "by": str(by)})

    def group_preamble_get(self, *, group_id: str) -> Dict[str, Any]:
        return self.call("group_preamble_get", {"group_id": str(group_id)})

    def group_preamble_set(self, *, group_id: str, content: str, by: str = "user") -> Dict[str, Any]:
        if not isinstance(content, str) or not content.strip():
            raise ValueError("group_preamble_set requires non-empty content")
        body = content
        if len(body.encode("utf-8")) > _MAX_GROUP_PREAMBLE_BYTES:
            raise ValueError("group_preamble_set content exceeds 512 KiB")
        return self.call(
            "group_preamble_set",
            {"group_id": str(group_id), "content": body, "by": str(by)},
        )

    def group_preamble_reset(self, *, group_id: str, confirm: str, by: str = "user") -> Dict[str, Any]:
        if str(confirm) != "preamble":
            raise ValueError("group_preamble_reset requires confirm='preamble'")
        return self.call(
            "group_preamble_reset",
            {"group_id": str(group_id), "confirm": "preamble", "by": str(by)},
        )

    def terminal_history(
        self,
        *,
        group_id: str,
        actor_id: str,
        before: Optional[int] = None,
        limit_bytes: Optional[int] = None,
        strip_ansi: Optional[bool] = None,
        compact: Optional[bool] = None,
        limit: Optional[int] = None,
        cursor: str = "",
        by: str = "user",
    ) -> Dict[str, Any]:
        cursor_before = int(cursor) if cursor.isdigit() else None
        return self.call(
            "terminal_history",
            _compact(
                {
                    "group_id": str(group_id),
                    "actor_id": str(actor_id),
                    "before": int(before) if before is not None else cursor_before,
                    "limit_bytes": int(limit_bytes) if limit_bytes is not None else (int(limit) if limit is not None else None),
                    "strip_ansi": strip_ansi,
                    "compact": compact,
                    "by": str(by),
                }
            ),
        )

    def terminal_since(
        self,
        *,
        group_id: str,
        actor_id: str,
        after: int,
        limit_bytes: Optional[int] = None,
        by: str = "user",
    ) -> Dict[str, Any]:
        return self.call(
            "terminal_since",
            _compact(
                {
                    "group_id": str(group_id),
                    "actor_id": str(actor_id),
                    "after": int(after),
                    "limit_bytes": int(limit_bytes) if limit_bytes is not None else None,
                    "by": str(by),
                }
            ),
        )

    def term_resize(self, *, group_id: str, actor_id: str, cols: int, rows: int) -> Dict[str, Any]:
        args = {"group_id": str(group_id), "actor_id": str(actor_id), "cols": int(cols), "rows": int(rows)}
        try:
            return self.call("term_resize", args)
        except DaemonAPIError as error:
            if error.code != "unknown_op":
                raise
        # Rust CCCC builds prior to contract parity used this legacy alias.
        legacy = self.call("terminal_resize", args)
        return {
            "group_id": str(group_id),
            "actor_id": str(actor_id),
            "cols": int(legacy.get("cols") or cols),
            "rows": int(legacy.get("rows") or rows),
        }

    def im_bind_chat(
        self,
        *,
        group_id: str,
        key: str,
    ) -> Dict[str, Any]:
        return self.call("im_bind_chat", {"group_id": str(group_id), "key": str(key)})

    def im_list_authorized(self, *, group_id: str) -> Dict[str, Any]:
        return self.call("im_list_authorized", {"group_id": str(group_id)})

    def im_list_pending(self, *, group_id: str) -> Dict[str, Any]:
        return self.call("im_list_pending", {"group_id": str(group_id)})

    def im_reject_pending(self, *, group_id: str, key: str) -> Dict[str, Any]:
        return self.call("im_reject_pending", {"group_id": str(group_id), "key": str(key)})

    def im_revoke_chat(
        self,
        *,
        group_id: str,
        chat_id: str,
        thread_id: Optional[Union[int, str]] = None,
    ) -> Dict[str, Any]:
        return self.call(
            "im_revoke_chat",
            _compact(
                {
                    "group_id": str(group_id),
                    "chat_id": str(chat_id),
                    "thread_id": thread_id,
                }
            ),
        )

    def remote_access_state(self, *, by: str = "user") -> Dict[str, Any]:
        return self.call("remote_access_state", {"by": str(by)})

    def remote_access_configure(
        self,
        *,
        provider: Optional[str] = None,
        mode: Optional[str] = None,
        require_access_token: Optional[bool] = None,
        web_host: Optional[str] = None,
        web_port: Optional[int] = None,
        web_public_url: Optional[str] = None,
        by: str = "user",
    ) -> Dict[str, Any]:
        return self.call(
            "remote_access_configure",
            _compact(
                {
                    "by": str(by),
                    "provider": provider,
                    "mode": mode,
                    "require_access_token": require_access_token,
                    "web_host": web_host,
                    "web_port": int(web_port) if web_port is not None else None,
                    "web_public_url": web_public_url,
                }
            ),
        )

    def remote_access_start(self, *, by: str = "user") -> Dict[str, Any]:
        return self.call("remote_access_start", {"by": str(by)})

    def remote_access_stop(self, *, by: str = "user") -> Dict[str, Any]:
        return self.call("remote_access_stop", {"by": str(by)})

File: ts/src/client_0430_admin_ops.ts
import { Buffer } from 'node:buffer';
import {
  compactRecord,
  type BasicGroupActorOptions,
  type CCCC0430Client,
} from './client_0430_shared.js';
import { DaemonAPIError } from './errors.js';
import type {
  GroupCopyExportOptions,
  GroupPreambleResetOptions,
  GroupPreambleSetOptions,
  RemoteAccessConfigureOptions,
  RemoteAccessOptions,
} from './types.js';

const MAX_GROUP_PREAMBLE_BYTES = 512 * 1024;

export interface CCCC0430AdminOps {
  actorNewSession(groupId: string, actorId: string, by?: string): Promise<Record<string, unknown>>;
  actorNewSession(options: BasicGroupActorOptions): Promise<Record<string, unknown>>;
  groupCopyExportFile(options: GroupCopyExportOptions): Promise<Record<string, unknown>>;
  groupPreambleGet(options: { groupId: string }): Promise<Record<string, unknown>>;
  groupPreambleSet(options: GroupPreambleSetOptions): Promise<Record<string, unknown>>;
  groupPreambleReset(options: GroupPreambleResetOptions): Promise<Record<string, unknown>>;
  terminalHistory(options: BasicGroupActorOptions & {
    before?: number;
    limitBytes?: number;
    stripAnsi?: boolean;
    compact?: boolean;
    /** @deprecated Use limitBytes. */
    limit?: number;
    /** @deprecated Use before. Numeric strings are accepted. */
    cursor?: string;
  }): Promise<Record<string, unknown>>;
  terminalSince(options: BasicGroupActorOptions & { after: number; limitBytes?: number }): Promise<Record<string, unknown>>;
  termResize(options: BasicGroupActorOptions & { cols: number; rows: number }): Promise<Record<string, unknown>>;
  imBindChat(options: { groupId: string; key: string }): Promise<Record<string, unknown>>;
  imListAuthorized(options: { groupId: string }): Promise<Record<string, unknown>>;
  imListPending(options: { groupId: string }): Promise<Record<string, unknown>>;
  imRejectPending(options: { groupId: string; key: string }): Promise<Record<string, unknown>>;
  imRevokeChat(options: { groupId: string; chatId: string; threadId?: number | string }): Promise<Record<string, unknown>>;
  remoteAccessState(options?: RemoteAccessOptions): Promise<Record<string, unknown>>;
  remoteAccessConfigure(options: RemoteAccessConfigureOptions): Promise<Record<string, unknown>>;
  remoteAccessStart(options?: RemoteAccessOptions): Promise<Record<string, unknown>>;
  remoteAccessStop(options?: RemoteAccessOptions): Promise<Record<string, unknown>>;
}

const adminOps: CCCC0430AdminOps & ThisType<CCCC0430Client> = {
  async actorNewSession(
    optionsOrGroupId: BasicGroupActorOptions | string,
    legacyActorId?: string,
    legacyBy: string = 'user',
  ) {
    const options: BasicGroupActorOptions = typeof optionsOrGroupId === 'string'
      ? { groupId: optionsOrGroupId, actorId: String(legacyActorId ?? ''), by: legacyBy }
      : optionsOrGroupId;
    return this.call('actor_new_session', {
      group_id: options.groupId,
      actor_id: options.actorId,
      by: options.by ?? 'user',
    });
  },

  async groupCopyExportFile(options) {
    return this.call('group_copy_export_file', compactRecord({
      group_id: options.groupId,
      by: options.by ?? 'user',
    }));
  },

  async groupPreambleGet(options) {
    return this.call('group_preamble_get', { group_id: options.groupId });
  },

  async groupPreambleSet(options) {
    if (typeof options.content !== 'string' || options.content.trim().length === 0) {
      throw new DaemonAPIError('invalid_args', 'groupPreambleSet requires non-empty content', {});
    }
    if (Buffer.byteLength(options.content, 'utf8') > MAX_GROUP_PREAMBLE_BYTES) {
      throw new DaemonAPIError('invalid_args', 'groupPreambleSet content exceeds 512 KiB', {});
    }
    return this.call('group_preamble_set', {
      group_id: options.groupId,
      content: options.content,
      by: options.by ?? 'user',
    });
  },

  async groupPreambleReset(options) {
    if (options.confirm !== 'preamble') {
      throw new DaemonAPIError('invalid_args', "groupPreambleReset requires confirm='preamble'", {});
    }
    return this.call('group_preamble_reset', {
      group_id: options.groupId,
      confirm: 'preamble',
      by: options.by ?? 'user',
    });
  },

  async terminalHistory(options) {
    const cursorBefore = options.cursor === undefined ? undefined : Number(options.cursor);
    return this.call('terminal_history', compactRecord({
      group_id: options.groupId,
      actor_id: options.actorId,
      before: options.before ?? (Number.isSafeInteger(cursorBefore) ? cursorBefore : undefined),
      limit_bytes: options.limitBytes ?? options.limit,
      strip_ansi: options.stripAnsi,
      compact: options.compact,
      by: options.by ?? 'user',
    }));
  },

  async terminalSince(options) {
    return this.call('terminal_since', compactRecord({
      group_id: options.groupId,
      actor_id: options.actorId,
      after: options.after,
      limit_bytes: options.limitBytes,
      by: options.by ?? 'user',
    }));
  },

  async termResize(options) {
    const args = {
      group_id: options.groupId,
      actor_id: options.actorId,
      cols: options.cols,
      rows: options.rows,
    };
    try {
      return await this.call('term_resize', args);
    } catch (error) {
      if (!(error instanceof DaemonAPIError) || error.code !== 'unknown_op') {
        throw error;
      }
    }
    // Rust CCCC builds prior to contract parity used this legacy alias.
    const legacy = await this.call('terminal_resize', args);
    return {
      group_id: options.groupId,
      actor_id: options.actorId,
      cols: typeof legacy['cols'] === 'number' ? legacy['cols'] : options.cols,
      rows: typeof legacy['rows'] === 'number' ? legacy['rows'] : options.rows,
    };
  },

  async imBindChat(options) {
    return this.call('im_bind_chat', {
      group_id: options.groupId,
      key: options.key,
    });
  },

  async imListAuthorized(options) {
    return this.call('im_list_authorized', { group_id: options.groupId });
  },

  async imListPending(options) {
    return this.call('im_list_pending', { group_id: options.groupId });
  },

  async imRejectPending(options) {
    return this.call('im_reject_pending', {
      group_id: options.groupId,
      key: options.key,
    });
  },

  async imRevokeChat(options) {
    return this.call('im_revoke_chat', compactRecord({
      group_id: options.groupId,
      chat_id: options.chatId,
      thread_id: options.threadId,
    }));
  },

  async remoteAccessState(options = {}) {
    return this.call('remote_access_state', { by: options.by ?? 'user' });
  },

  async remoteAccessConfigure(options) {
    return this.call('remote_access_configure', compactRecord({
      by: options.by ?? 'user',
      provider: options.provider,
      mode: options.mode,
      require_access_token: options.requireAccessToken,
      web_host: options.webHost,
      web_port: options.webPort,
      web_public_url: options.webPublicUrl,
    }));
  },

  async remoteAccessStart(options = {}) {
    return this.call('remote_access_start', { by: options.by ?? 'user' });
  },

  async remoteAccessStop(options = {}) {
    return this.call('remote_access_stop', { by: options.by ?? 'user' });
  },

};

export function installCCCC0430AdminOps(proto: CCCC0430Client & Partial<CCCC0430AdminOps>): void {
  Object.assign(proto, adminOps);
}
